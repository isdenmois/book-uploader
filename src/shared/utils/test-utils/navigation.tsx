import React from 'react'
import { NavigationContext } from '@react-navigation/native'

export const mockNavigation = () => {
  const navContext = {
    isFocused: () => true,
    goBack: jest.fn(),
    addListener: jest.fn(() => jest.fn()),
  }

  const NavigationWrapper = ({ children }) => (
    <NavigationContext.Provider value={navContext as any}>{children}</NavigationContext.Provider>
  )

  return { navContext, NavigationWrapper }
}
