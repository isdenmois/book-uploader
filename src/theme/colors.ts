export const primary = '#000';
export const secondary = '#eee';
export const invert = '#fff';

export const green = '#8BC34A';
export const red = '#f44336';
export const darkBlue = '#007ac1';
export const blue = '#2196f3';
export const lightBlue = '#67daff';

export const Background = '#F9FAFE';
export const SearchBackground = '#B8D5F9';
export const UploadBackground = '#C2B9F9';
export const ProfileBackground = '#FFAD99';
export const TabsBackground = 'white';

export const SearchText = '#092D5D';
export const UploadText = '#15085E';
export const ProfileText = '#661400';
export const TabText = '#121417';

export const SearchSelected = '#5396EE';
export const UploadSelected = '#6851F0';
export const ProfileSelected = '#FF6842';

export const Secondary = '#787D87';
export const Search = '#8C93A3';
export const Input = 'white';
export const DeleteButton = '#9BA8CA';
export const Success = '#48B755';
export const Error = '#FD4463';
