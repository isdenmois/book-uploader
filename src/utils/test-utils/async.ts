import { renderHook, RenderHookOptions, RenderHookResult } from '@testing-library/react-hooks';
import { act } from '@testing-library/react-native';
import { create, ReactTestRenderer } from 'react-test-renderer';

/*global jest */

type FunctionPropertyNames<T> = { [K in keyof T]: T[K] extends (...args: any[]) => any ? K : never }[keyof T] & string;

export const mockPromise = <T extends {}, M extends FunctionPropertyNames<Required<T>>>(object: T, method: M) => {
  let resolvePromise;
  const promise: any = new Promise(resolve => (resolvePromise = resolve));

  if (method in object) {
    jest.spyOn(object, method).mockReturnValue(promise);
  } else {
    (object[method] as any) = () => promise;
  }

  const resolve = value => act(async () => resolvePromise(value));

  return [resolve] as const;
};

export const mockPromiseValue = (obj, prop, value) => {
  const promise = Promise.resolve(value);

  if (prop in obj) {
    jest.spyOn(obj, prop).mockReturnValue(promise);
  } else {
    obj[prop] = () => promise;
  }
};

export const createAsync = async (elem): Promise<ReactTestRenderer> => {
  let result: ReactTestRenderer;

  await act(async () => (result = create(elem)));

  return result;
};

export const immediate = () => new Promise(resolve => setImmediate(resolve));

export async function renderHookAsync<P, R>(
  callback: (props: P) => R,
  options?: RenderHookOptions<P>,
): Promise<RenderHookResult<P, R>> {
  let result: RenderHookResult<P, R>;

  await act(async () => (result = renderHook(callback, options)));

  return result;
}

export function mock<T extends {}, M extends FunctionPropertyNames<Required<T>>>(object: T, method: M) {
  if (method in object) {
    return jest.spyOn(object, method);
  }

  return ((object[method] as any) = () => jest.fn());
}
