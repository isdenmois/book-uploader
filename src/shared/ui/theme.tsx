import React, { FC } from 'react'
import { TouchableOpacity, TouchableOpacityProps, useColorScheme } from 'react-native'
import {
  ThemeProvider as ReThemeProvider,
  createBox,
  createText,
  createTheme,
  useTheme as useThemeRN,
} from '@shopify/restyle'

import * as Colors from './colors'

export const theme = createTheme({
  spacing: {
    0.25: 2,
    0.5: 4,
    1: 8,
    2: 16,
    3: 24,
    4: 32,
    5: 40,
    6: 48,
  },
  colors: Colors.light,
  breakpoints: {},
  textVariants: {
    defaults: {
      fontSize: 16,
    },
    chip: {
      fontSize: 14,
      lineHeight: 18,
    },
    button_primary: {
      fontSize: 16,
      lineHeight: 22,
      color: 'invertedText',
    },
  },
  buttonVariants: {
    primary: {
      backgroundColor: 'profileSelected',
    },
  },
})

export type Theme = typeof theme

const darkTheme: Theme = {
  ...theme,
  colors: {
    ...theme.colors,
    ...Colors.dark,
  },
}

export const Box = createBox<Theme>()
export const Text = createText<Theme>()
export const TouchableBox = createBox<Theme, TouchableOpacityProps & { children: any }>(TouchableOpacity)

export const ThemeProvider: FC = ({ children }) => {
  const currentMode = useColorScheme()

  return <ReThemeProvider theme={currentMode === 'dark' ? darkTheme : theme}>{children}</ReThemeProvider>
}

export const useTheme = () => useThemeRN<Theme>()
