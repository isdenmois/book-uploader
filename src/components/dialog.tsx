import { useNavigation } from '@react-navigation/native';
import React from 'react';
import { TouchableWithoutFeedback, StyleSheet, View, ViewStyle } from 'react-native';
import * as colors from 'theme/colors';

export function Dialog({ children }) {
  const navigation = useNavigation();

  return (
    <View style={s.container}>
      <TouchableWithoutFeedback onPress={() => navigation.goBack()}>
        <View style={s.overlay} />
      </TouchableWithoutFeedback>

      <View style={s.content}>{children}</View>
    </View>
  );
}

const s = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    overflow: 'hidden',
  } as ViewStyle,
  overlay: {
    backgroundColor: '#0007',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  content: {
    backgroundColor: colors.ModalBackground,
    borderRadius: 20,
    overflow: 'hidden',
  } as ViewStyle,
});
