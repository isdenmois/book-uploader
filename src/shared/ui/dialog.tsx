import React from 'react'
import { StyleSheet, TouchableWithoutFeedback, View } from 'react-native'
import { useNavigation } from '@react-navigation/native'
import { Box } from './theme'

export function Dialog({ children }) {
  const navigation = useNavigation()

  return (
    <View style={s.container}>
      <TouchableWithoutFeedback onPress={() => navigation.goBack()} testID='dialogOverlay'>
        <View style={s.overlay} />
      </TouchableWithoutFeedback>

      <Box style={s.content} backgroundColor='modalBackground'>
        {children}
      </Box>
    </View>
  )
}

const s = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    position: 'relative',
    overflow: 'hidden',
  },
  overlay: {
    backgroundColor: '#0007',
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  content: {
    borderRadius: 20,
    overflow: 'hidden',
    margin: 48,
  },
})
