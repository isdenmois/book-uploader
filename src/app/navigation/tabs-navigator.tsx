import React, { useCallback } from 'react'
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs'
import { useNavigation } from '@react-navigation/native'

import { AccountIcon, SearchIcon, tabBar, UploadIcon } from 'shared/ui'
import { useDeepLink } from 'shared/utils'
import { setQuery } from 'entities/search-filters'

import { SearchScreen } from 'screens/search.screen'
import { UploadScreen } from 'screens/upload.screen'
import { ProfileScreen } from 'screens/profile.screen'

const Tab = createBottomTabNavigator()

const SCREENS_OPTIONS: Record<string, any> = {
  search: { color: 'searchBackground', textColor: 'searchText', icon: SearchIcon },
  upload: { color: 'uploadBackground', textColor: 'uploadText', icon: UploadIcon },
  profile: { color: 'profileBackground', textColor: 'profileText', icon: AccountIcon },
}

export function TabsNavigator() {
  useInitialQuery()

  return (
    <Tab.Navigator tabBar={tabBar}>
      <Tab.Screen name='Upload' component={UploadScreen} options={SCREENS_OPTIONS.upload} />
      <Tab.Screen name='Search' component={SearchScreen} options={SCREENS_OPTIONS.search} />
      <Tab.Screen name='Profile' component={ProfileScreen} options={SCREENS_OPTIONS.profile} />
    </Tab.Navigator>
  )
}

function useInitialQuery() {
  const navigation = useNavigation()
  const onLink = useCallback(link => {
    if (link) {
      link = link.replace('booksearch://', '')
      navigation.navigate('Search')
    }

    setQuery(link || '')
  }, [])

  useDeepLink(onLink)
}
