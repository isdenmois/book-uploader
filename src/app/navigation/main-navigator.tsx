import React from 'react'
import { createStackNavigator } from '@react-navigation/stack'

import { ScanScreen } from 'screens/scan'
import { DownloadModal } from 'screens/download.modal'

import { TabsNavigator } from './tabs-navigator'

const Stack = createStackNavigator()

export function MainNavigator() {
  return (
    <Stack.Navigator>
      <Stack.Group screenOptions={{ detachPreviousScreen: false }}>
        <Stack.Screen name='Tabs' component={TabsNavigator} options={{ headerShown: false }} />
      </Stack.Group>

      <Stack.Group screenOptions={modalOptions}>
        <Stack.Screen name='Scan' component={ScanScreen} />
        <Stack.Screen name='Download' component={DownloadModal} />
      </Stack.Group>
    </Stack.Navigator>
  )
}

const modalOptions: any = {
  detachPreviousScreen: false,
  presentation: 'modal',
  headerShown: false,
  cardStyle: {
    backgroundColor: 'transparent',
  },
  cardOverlayEnabled: false,
  cardShadowEnabled: false,

  cardStyleInterpolator: ({ current: { progress } }) => ({
    cardStyle: {
      opacity: progress.interpolate({
        inputRange: [0, 0.5, 0.9, 1],
        outputRange: [0, 0.1, 0.3, 1],
        extrapolate: 'clamp',
      }),
    },
  }),
}
