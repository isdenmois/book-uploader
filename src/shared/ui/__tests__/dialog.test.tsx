import React from 'react'
import { act, fireEvent, render } from '@testing-library/react-native'
import { mockNavigation } from 'shared/utils/test-utils/navigation'
import { ThemeProvider } from '../theme'
import { Dialog } from '../dialog'

test('Dialog', async () => {
  const { NavigationWrapper, navContext } = mockNavigation()
  const wrapper = ({ children }) => (
    <ThemeProvider>
      <NavigationWrapper>{children}</NavigationWrapper>
    </ThemeProvider>
  )
  const { toJSON, getByTestId, unmount } = render(<Dialog>test</Dialog>, { wrapper })

  expect(toJSON()).toMatchSnapshot()

  fireEvent.press(getByTestId('dialogOverlay'))

  expect(navContext.goBack).toHaveBeenCalled()

  await act(async () => {
    unmount()
  })
})
