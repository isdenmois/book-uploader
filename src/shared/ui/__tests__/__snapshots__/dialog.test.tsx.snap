// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#0007",
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="dialogOverlay"
  />
  <View
    style={
      Array [
        Object {
          "backgroundColor": "white",
          "display": "flex",
        },
        Object {
          "borderRadius": 20,
          "margin": 48,
          "overflow": "hidden",
        },
      ]
    }
  >
    test
  </View>
</View>
`;
